modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorHighlightSubFetchPhase.java

```diff
@@ -99,15 +98,18 @@ public void hitsExecute(SearchContext context, SearchHit[] hits) throws IOExcept
                     for (Object matchedSlot : field.getValues()) {
                         int slot = (int) matchedSlot;
                         BytesReference document = percolateQuery.getDocuments().get(slot);
-                        SubSearchContext subSearchContext =
-                            createSubSearchContext(context, percolatorLeafReaderContext, document, slot);
-                        subSearchContext.parsedQuery(new ParsedQuery(query));
+                        SearchContextHighlight highlight = new SearchContextHighlight(context.highlight().fields());
+                        // Enforce highlighting by source, because MemoryIndex doesn't support stored fields.
+                        highlight.globalForceSource(true);
+                        QueryShardContext shardContext = new QueryShardContext(context.getQueryShardContext());
+                        shardContext.lookup().source().setSegmentAndDocument(percolatorLeafReaderContext, segmentDocId);
+                        shardContext.lookup().source().setSource(document);
                         hitContext.reset(
                             new SearchHit(slot, "unknown", Collections.emptyMap()),
                             percolatorLeafReaderContext, slot, percolatorIndexSearcher
                         );
                         hitContext.cache().clear();
-                        highlightPhase.hitExecute(subSearchContext, hitContext);
+                        highlightPhase.hitExecute(context.shardTarget(), context.getQueryShardContext(), query, highlight, hitContext);
```

Comment:
Should this be passing `shardContext` rather than `context.getQueryShardContext()`?